plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -169,22 +211,117 @@ public ConnectorPageSource createPageSource(
     {
         IcebergSplit split = (IcebergSplit) connectorSplit;
         IcebergTableHandle table = (IcebergTableHandle) connectorTable;
-
-        List<IcebergColumnHandle> icebergColumns = columns.stream()
+        Table icebergTable = table.getTable();
+        FileScanTask task = split.getTask();
+        List<IcebergColumnHandle> updateColumns = table.getUpdateColumns();
+        List<IcebergColumnHandle> queriedColumns = columns.stream()
                 .map(IcebergColumnHandle.class::cast)
                 .collect(toImmutableList());
 
         Map<Integer, Optional<String>> partitionKeys = split.getPartitionKeys();
+        Optional<StructLike> partition = task.spec().isUnpartitioned() ? Optional.empty() : Optional.of(task.file().partition());
+        Optional<PartitionData> partitionDataForWrite = coercePartitionData(icebergTable.spec(), task.spec(), partition);
 
-        List<IcebergColumnHandle> regularColumns = columns.stream()
-                .map(IcebergColumnHandle.class::cast)
-                .filter(column -> !partitionKeys.containsKey(column.getId()))
-                .collect(toImmutableList());
         TupleDomain<IcebergColumnHandle> effectivePredicate = table.getUnenforcedPredicate()
                 .intersect(dynamicFilter.getCurrentPredicate().transformKeys(IcebergColumnHandle.class::cast))
                 .simplify(ICEBERG_DOMAIN_COMPACTION_THRESHOLD);
 
+        // construct the columns that needs to be read by the file reader
+        List<Integer> fileReadColumnIds = new ArrayList<>();
+        List<IcebergColumnHandle> fileReadColumns = new ArrayList<>();
+        List<Boolean> isRowPositionChannel = new ArrayList<>();
+
+        // 1. non-partition queried columns
+        Block[] queriedColumnPrefillValues = new Block[queriedColumns.size()];
+        int[] queriedColumnFileReadChannels = new int[queriedColumns.size()];
+        boolean isDeleteOrUpdateQuery = false;
+        for (int idx = 0; idx < queriedColumns.size(); idx++) {
+            IcebergColumnHandle column = queriedColumns.get(idx);
+            if (column.isTrinoRowIdColumn()) {
```

Comment:
Should we be more explicit and encode fact that we are doing DELETE/UPDATE in `IcebergTableHandle` instead of inferring it from list of columns?